delta_shooter/models.py

```diff
@@ -0,0 +1,268 @@
+        """Fudge factor stops bullets skipping over objects."""
+
+        fudge_factor = 1.5
+        distance = self.position.distance_to(other_obj.position)
+        return distance < (self.radius * fudge_factor) + (other_obj.radius * fudge_factor)
+
+
+class Spaceship(GameObject):
+    ANGLE_TURN = 15
+    ACCELERATION = 0.1
+    BULLET_SPEED = 60
+    NUM_BULLETS = 2  # Limit the number on the screen at one time
+
+    def __init__(
+        self,
+        starting_position: Tuple[int, int],
+        starting_orientation: Vector2,
+        player: int,
+        graphical: bool = True,
+    ) -> None:
+
+        self.starting_position = starting_position
+        self.starting_orientation = starting_orientation
+        self.graphical = graphical
+        self.player = player
+        self.dead = False
+        self.name = "spaceship"
+
+        if self.graphical:
+            super().__init__(
+                starting_position, load_sprite(f"spaceship_player{player}"), Vector2(0)
+            )
+            self.laser_sound = load_sound("laser")
+        else:
+            super().__init__(starting_position, DummyShip(), Vector2(0))
+            self.laser_sound = DummySound()
+
+        # self.radius *= 1.5
+        self.reset()
+
+    def reset(self) -> None:
+        self.set_position(self.starting_position)
+        self.set_orientation(self.starting_orientation)
+
+        self.bullets: List[Bullet] = []
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Spaceship):
+            raise NotImplemented
+        return self.player == other.player
+
+    def rotate(self, clockwise: bool = True) -> None:
+        sign = 1 if clockwise else -1
+        angle = self.ANGLE_TURN * sign
+        self.direction.rotate_ip(angle)
+
+    def accelerate(self) -> None:
+        self.velocity += self.direction * self.ACCELERATION
+
+    def move_forward(self) -> None:
+        distance = self.radius
+        new_position = self.position + self.direction * distance
+        for barrier in BARRIERS:
+            if barrier.hit_barrier(self.position, new_position, self.radius):
+                return
+        self.position = new_position
+
+    def draw(self, surface: pygame.Surface) -> None:
+        angle = self.direction.angle_to(UP)
+        rotated_surface = rotozoom(self.sprite, angle, 1.0)
+        rotated_surface_size = Vector2(rotated_surface.get_size())
+        blit_position = self.position - rotated_surface_size * 0.5
+        surface.blit(rotated_surface, blit_position)
+
+    def shoot(self) -> None:
+        # Limit number of bullets
+        if len(self.bullets) == self.NUM_BULLETS:
+            return
+        bullet_velocity = self.direction * self.BULLET_SPEED + self.velocity
+        bullet = Bullet(self.position, bullet_velocity, self.graphical)
+        self.bullets.append(bullet)
+        self.laser_sound.play()
+
+
+class Bullet(GameObject):
+    def __init__(self, position, velocity, graphical):
```

Comment:
Types please - they're missing in a couple of other places in this file, so would be good to tighten these up :smile: